cmd/panfs-multipart.go

```diff
@@ -73,68 +106,121 @@ func (fs *PANFSObjects) decodePartFile(name string) (partNumber int, etag string
+	// Since we append sequentially nextPartNumber will always be len(fsParts.Appended)+1
+	initialAppendedLen := len(fsParts.Parts)
+	nextPartNumber := initialAppendedLen + 1
 
-	for _, entry := range entries {
-		if entry == fs.metaJSONFile {
-			continue
-		}
-		partNumber, etag, actualSize, err := fs.decodePartFile(entry)
+	for {
+		entries, err := readDir(uploadIDDir)
 		if err != nil {
-			// Skip part files whose name don't match expected format. These could be backend filesystem specific files.
-			continue
-		}
-		if partNumber < nextPartNumber {
-			// Part already appended.
-			continue
-		}
-		if partNumber > nextPartNumber {
-			// Required part number is not yet uploaded.
+			logger.GetReqInfo(ctx).AppendTags("uploadIDDir", uploadIDDir)
+			logger.LogIf(ctx, err)
 			return
 		}
+		sort.Strings(entries)
 
-		partPath := pathJoin(uploadIDDir, entry)
-		err = xioutil.AppendFile(file.filePath, partPath, globalFSOSync)
+		for _, entry := range entries {
+			if entry == fs.metaJSONFile {
+				continue
+			}
+			partNumber, etag, actualSize, err := fs.decodePartFile(entry)
+			if err != nil {
+				// Skip part files whose name don't match expected format. These could be backend filesystem specific files.
+				continue
+			}
+			if partNumber < nextPartNumber {
+				// Part already appended.
+				continue
+			}
+			if partNumber > nextPartNumber {
+				// Required part number is not yet uploaded.
+				break
+			}
+
+			partPath := pathJoin(uploadIDDir, entry)
+			err = xioutil.AppendFile(file.filePath, partPath, globalFSOSync)
+			if err != nil {
+				reqInfo := logger.GetReqInfo(ctx).AppendTags("partPath", partPath)
+				reqInfo.AppendTags("filepath", file.filePath)
+				logger.LogIf(ctx, err)
+				return
+			}
+
+			partInfo := PartInfo{PartNumber: partNumber, ETag: etag, ActualSize: actualSize}
+			fsParts.Parts = append(fsParts.Parts, partInfo)
+			file.parts = append(file.parts, partInfo)
+			nextPartNumber++
+		}
+
+		uploadIDDirStat, err := fsStatDir(ctx, uploadIDDir)
 		if err != nil {
-			reqInfo := logger.GetReqInfo(ctx).AppendTags("partPath", partPath)
-			reqInfo.AppendTags("filepath", file.filePath)
-			logger.LogIf(ctx, err)
-			return
+			// Do not return at this step - some parts may be appended to the file but not added to meta
+			break
+		}
+		if uploadIDDirStat.ModTime() == prevUploadIDDirStat.ModTime() {
+			break
 		}
+		prevUploadIDDirStat = uploadIDDirStat
+	}
+
+	// No parts appended - just return. No need to update meta file
+	if initialAppendedLen == len(fsParts.Parts) {
+		return
+	}
 
-		file.parts = append(file.parts, PartInfo{PartNumber: partNumber, ETag: etag, ActualSize: actualSize})
-		nextPartNumber++
+	// Write info about appended parts
+	panfsPartsBytes, err := json.Marshal(fsParts)
+	if err != nil {
+		logger.LogIf(ctx, err)
+		return
+	}
+
+	// Update info about uploaded parts
+	panfsPartsTmpPath := pathJoin(bucketPath, panfsS3TmpDir, mustGetUUID())
+	defer fsRemoveFile(ctx, panfsPartsTmpPath)
```

Comment:
I'd enqueue fsRemoveFile only if write or rename fails. Just to avoid redundant fs operations, although code is much cleaner now.